### PR TITLE
Reuse optimizer throughout training session

### PR DESCRIPTION
**Pull request by adamantivm**
As pointed out by @muralx

### CODE REVIEW

File: deep_quoridor/src/agents/alphazero/alphazero.py [hunk 1]
```diff
@@ -91,6 +91,8 @@ def __init__(
         self.action_encoder = ActionEncoder(board_size)
         self.evaluator = NNEvaluator(self.action_encoder, self.device)
         self.mcts = MCTS(params.mcts_n, params.mcts_ucb_c, self.evaluator)
+        if params.training_mode:
+            self.evaluator.train_prepare(params.learning_rate, params.batch_size, params.optimizer_iterations)
```
Comment (muralx) [inline deep_quoridor/src/agents/alphazero/alphazero.py:95]: nit: If you want to enable L2, you can add the param weight decay just to have it ready.

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py [hunk 2]
```diff
@@ -114,12 +114,19 @@ def game_to_input_array(self, game: Quoridor) -> tuple[torch.FloatTensor, bool]:
...
         return input_array
 
-    def train_network(self, replay_buffer, learning_rate, batch_size, optimizer_iterations):
-        optimizer = torch.optim.Adam(self.network.parameters(), lr=learning_rate)
+    def train_prepare(self, learning_rate, batch_size, optimizer_iterations):
+        assert not hasattr(self, "optimizer") or self.optimizer is None, "train_prepare should be called only once"
 
-        for _ in range(optimizer_iterations):
+        self.batch_size = batch_size
+        self.optimizer_iterations = optimizer_iterations
```
Comment (muralx) [inline deep_quoridor/src/agents/alphazero/nn_evaluator.py:121]: nit: batches_per_train_iteration?

**Review by muralx — APPROVED** [review]

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py (hunk 2 above)
Comment (adamantivm) [reply deep_quoridor/src/agents/alphazero/nn_evaluator.py]: done
